RHIDP-7560 - Install and configure plugins from Extensions page (#1281)
redhat-developer/red-hat-developers-documentation-rhdh: modules/dynamic-plugins/proc-extensions-managing.adoc

[id="rhdh-extensions-plugins-managing_{context}"]
= Managing plugins by using Extensions

[WARNING]

> Installation and configuration of plugins by using Extensions will only work in development environments. This feature is not supported in production environments.

Author: This content has been removed.

// I’m going to walk you through the end-to-end flow of how plugins can be installed in a development environment using the Extensions feature in the Red Hat Developer Hub.
// [NOTE]
// This installation feature is currently only meant for development environments. It is not supported in production environments yet.

// Plugin configuration and installation
// When installing a plugin, the configuration that is used during installing is saved to a dynamic plugins cache. This ensures that the same configuration is available when editing or re-enabling the plugin. The persist the cache across pod restarts, a persistent volume claim must be created and configured in the Helm chart.

.Prerequisites
* You added a custom {product} ({product-very-short}) application configuration, and have sufficient permissions to modify it.
* You have configured {product-very-short} to install dynamic plugins.
* You have created a persistent volume (PVC) for the dynamic plugin cache.

> * You have a user with required permissions [TBC]

Author: This content has been removed.

.Procedure

> //. Create a PVC with the name `dynamic-plugins-route`. This will be used in the Helm chart values when deploying the {product-very-short}.

Author: This content has been updated and included in _proc-extensions-enabling-plugins-installation.adoc_

////
. Create the PVC and save it to a file, such as `pvc.yaml`. For example:
+
.`pvc.yaml` fragment
[source,yaml]
----
kind: PersistentVolumeClaim
apiVersion: v1
metadata:
  name: dynamic-plugins-root
spec:
  accessModes:
    - ReadWriteOnce
  resources:
    requests:
      storage: 5Gi
----

. To apply this PVC to your cluster, run the following command:
+
[source,yaml]
----
oc apply -f pvc.yaml
----

. Configure the Helm chart to replace the ephermal volume claim with the PVC. For example:
+
[source,yaml]
----
upstream:
  backstage:
    extraVolumes:
      - name: dynamic-plugins-root
        persistentVolumeClaim:
          claimName: dynamic-plugins-root
----
////
// Now if i go to the extra volumes under upstream backstage app config, I’ll replace the ephermal volume claim with the PVC volume claim that I just created.

//. Create the Helm Release.

> . Create a marketplace configuration file and save it to `dynamic-plugins.marketplace.yaml`. For example:

Reviewer: I believe these are commented out, but maybe @Gerry-Forde can remove all big blocks of commented out content because it is harder to review what actually goes in.

+
.`dynamic-plugins.marketplace.yaml` fragment
[source,yaml]
----
plugins:
  - package: ./dynamic-plugins/dist/red-hat-developer-hub-backstage-plugin-marketplace-backend-dynamic
    disabled: false
    pluginConfig:
      extensions:
        installation:
          enabled: true
          saveToSingleFile:
            file: /opt/app-root/src/dynamic-plugins-root/dynamic-plugins.marketplace.yaml
  - package: ./dynamic-plugins/dist/red-hat-developer-hub-backstage-plugin-marketplace
    disabled: false
    pluginConfig:
      dynamicPlugins:
        frontend:
          red-hat-developer-hub.backstage-plugin-marketplace:
            appIcons:
              - name: marketplace
                importName: MarketplaceIcon
            dynamicRoutes:
              - path: /extensions/catalog
                importName: DynamicMarketplacePluginRouter
            mountPoints:

>               - mountPoint: application/provider

Author: This content has been updated and included in _proc-extensions-enabling-plugins-installation.adoc_

+
[source,yaml]
----
- package: ./dynamic-plugins/dist/red-hat-developer-hub-backstage-plugin-marketplace
  disabled: false
  pluginConfig:
     dynamicPlugins:
       frontend:
         red-hat-developer-hub.backstage-plugin-marketplace:
           appIcons:
             - name: marketplace
               importName: MarketplaceIcon
           dynamicRoutes:
             - path: /extensions/catalog
               importName: DynamicMarketplacePluginRouter
           mountPoints:
             - mountPoint: application/provider
               importName: InstallationContextProvider
             - mountPoint: internal.plugins/tab
               importName: DynamicMarketplacePluginContent
               config:
                 path: marketplace
                 title: Catalog
----

.Verification
Enable some plugin via UI that is defaultly disabled, restart and see the change.

////
Manually create a YAML file containing our new extensions backend configuration. The extensions plugin requires the installation to be set to true and a YAML file where your plugin configurations will be written to.

> I have created a local file named dynamic-plugins.marketplace.yaml and added the extensions configuration to it. I have specified the file path to be /opt/app-root/src/dynamic-plugins-root/dynamic-plugins.marketplace.yaml

Author: This content has been removed.

> Now, copy the file to the RHDH pod, fetch the pods running under my namespace. Copy the yaml file to the rhdh pod.

Author: This content has been removed.

Finally, I’ll upgrade the RHDH helm release to point to this dynamic plugins marketplace yaml. So under global dynamic includes, I’ll add /dynamic-plugins-root/dynamic plugins marketplace yaml.
Click upgrade

> In my dynamic plugins config map, I have added the extensions UI package configuration to include the newly added mount point:

Reviewer 2: ```suggestion
In your dynamic plugins config map, add the extensions UI package configuration to include the newly added mount point:
```

Author: This content has been removed.

> Once the RHDH pod is running, navigate to the extensions plugin.

Author: This content has been removed.

Select any plugin that you would like to install.
The install button is disabled because the current user does not have access to manage plugin configurations.
In RBAC, we need to create a role for this user to enable plugin installation.

Select extensions from the plugin dropdown, expand it to view the newly added permissions for the extensions plugin.
Select both and click next to create the role.
Refresh the application 
Now when I navigate to the extensions plugin and click on a plugin I see an Actions dropdown which means the plugin is preinstalled for me and I now have the access to edit or disable the plugin.

> If I disable the plugin, I’m notified to restart the backend to complete the action.

Reviewer 2: ```suggestion
If you disable the plugin, you are notified to restart the backend to complete the action.
```

Author: This content has been removed.

> Before I restart let’s install a plugin.

Author: This content has been removed.

> I’ll filter out the custom plugins, and select the Service Now plugin

Author: This content has been removed.

> I can see that the plugin is not already installed, so I’ll click on Install which takes me to the code editor. If I want to make any changes to this plugin configuration, I can. I’ll apply the default configuration for the front-end package:

Author: This content has been removed.

You can add a comment and click Install

> In this alert you can view all the plugins that require a backend restart

Author: This content has been removed.

Switch to my Openshift application and restart my RHDH pod (scale down/up).
Once the pod is running, switch back to your RHDH application and refresh the browser.

> Verification

Author: This content has been removed.

> Switch to the Adoption Insights plugin and I can see that it is disabled. The Service Now plugin shows the Actions dropdown with the option to edit and disable the plugin.

Author: This content has been removed.

> If I click on Edit, the configuration that I used to install the plugin has been loaded in the editor.

Reviewer 2: ```suggestion
If you click *Edit*, the configuration that you used to install the plugin is loaded in the editor.
```
